lib/extras/dec/exr.cc

```diff
@@ -127,52 +126,108 @@ class InMemoryIStream : public OpenEXR::IStream {
 
 bool CanDecodeEXR() { return true; }
 
+static std::string FindColorLayerPrefix(const OpenEXR::ChannelList& channels) {
+  // check if just R,G,B channels exist; use those if present
+  if (channels.findChannel("R") != nullptr &&
+      channels.findChannel("G") != nullptr &&
+      channels.findChannel("B") != nullptr) {
+    return "";
+  }
+
+  // check channels for presence of R,G,B with common prefix ("layer name"),
+  // use the first one that is present
+  for (OpenEXR::ChannelList::ConstIterator it = channels.begin(); it != channels.end(); ++it) {
+    std::string name = it.name();
+    size_t dotIndex = name.rfind('.');
+    if (dotIndex == std::string::npos)
+      continue;
+    std::string suffix = name.substr(dotIndex + 1);
+    if (suffix != "R" && suffix != "G" && suffix != "B")
+      continue;
+    std::string prefix = name.substr(0, dotIndex + 1);
+    if (channels.findChannel(prefix + "R") != nullptr &&
+        channels.findChannel(prefix + "G") != nullptr &&
+        channels.findChannel(prefix + "B") != nullptr) {
+      return prefix;
+    }
+  }
+
+  return "";
+}
+
 Status DecodeImageEXR(Span<const uint8_t> bytes, const ColorHints& color_hints,
                       PackedPixelFile* ppf,
                       const SizeConstraints* constraints) {
   InMemoryIStream is(bytes);
 
 #ifdef __EXCEPTIONS
-  std::unique_ptr<OpenEXR::RgbaInputFile> input_ptr;
+  std::unique_ptr<OpenEXR::InputFile> input_ptr;
   try {
-    input_ptr = jxl::make_unique<OpenEXR::RgbaInputFile>(is);
+    input_ptr = jxl::make_unique<OpenEXR::InputFile>(is);
   } catch (...) {
     // silently return false if it is not an EXR file
     return false;
   }
-  OpenEXR::RgbaInputFile& input = *input_ptr;
+  OpenEXR::InputFile& input = *input_ptr;
 #else
-  OpenEXR::RgbaInputFile input(is);
+  OpenEXR::InputFile input(is);
 #endif
 
-  if ((input.channels() & OpenEXR::RgbaChannels::WRITE_RGB) !=
-      OpenEXR::RgbaChannels::WRITE_RGB) {
-    return JXL_FAILURE("only RGB OpenEXR files are supported");
+  const OpenEXR::Header& header = input.header();
+  const OpenEXR::ChannelList& channels = header.channels();
+
+  // we don't support subsampled or UINT channels yet
```

Comment:
Maybe add a `// TODO:` comment here, since some cases of subsampling could be handled by setting the channel shift; this only works if `ch.xSampling == ch.ySampling` ∈ {2, 4, 8} but that probably does cover most of the cases of subsampling that occur in practice.